spring-batch-core/src/main/java/org/springframework/batch/core/job/builder/FlowBuilder.java

```diff
@@ -247,29 +248,32 @@ protected Flow flow() {
 	}
 
 	private void doNext(Object input) {
-		if (this.currentState == null) {
+		if (currentState == null) {
 			doStart(input);
+		} else {
+			State next = createState(input);
+			addTransition("COMPLETED", next);
+			addTransition("*", failedState);
+			currentState = next;
 		}
-		State next = createState(input);
-		addTransition("COMPLETED", next);
-		addTransition("*", failedState);
-		this.currentState = next;
```

Comment:
```java
		if (this.currentState == null) {
			doStart(input); // 1. currentState: input
		}

		State next = createState(input);
        // 2. input -> COMPLETED -> input -> COMPLETED -> input ..
		addTransition("COMPLETED", next); 
```

when `currentState == null`, I think we should not `addTransition("COMPLETED", next);` cause it makes infinite loop 🤔